### ISSUE TITLE
Why are there shadows on the face in the generated video？Apple M1 Pro use coreml provider

### ISSUE THREAD

**Issue by willstudy** [C0untFloyd/roop-unleashed#357]
### **Describe the bug**
Why are there shadows on the face in the generated video？Apple M1 Pro use **coreml provider**  & GFPGAN face enhancer

![cbd81f65-2b99-41fc-a2ab-811c8e72d29e](https://github.com/C0untFloyd/roop-unleashed/assets/12168193/1bf34c04-6c26-4fd6-b409-a555e4f2bc24)

But it's no shadows on the face if choose **cpu provider** & GFPGAN face enhancer !!

Thanks for help

**Details**
- Apple M1 Pro
- MEM: 32G 
- CPU: 10 (8 performance and 2 energy efficiency)
- GPU: 16 cores

**Which version of roop unleashed are you using?**
latest commit id 4bc0fc3fe7bba5d9a9d7d8fdd1fb97b5bb2e9754

**Screenshots**

Coreml does not have an acceleration effect and is as slow as the CPU's effect, only 1.5frames / s. 

<img width="1635" alt="截屏2023-11-28 21 25 36" src="https://github.com/C0untFloyd/roop-unleashed/assets/12168193/28d4e3a5-bb81-4e08-89a8-3dd227e59816">

**Comment by lysxelapsed** [comment]
Regarding the shadow: Yes, the developer (still on sick leave) ported the models to the onnx backend (which handles the GPU acceleration) - A good idea, but it's still buggy with some models. Codeformer seems to have a similar problem. Oil3 posted a solution at the end of #331 (couldn't test it yet) - But that's not the issue with GFPGAN, since "name = None" is already in the Enhance_GFPGAN.py file. It _could_ be related to your performance issue, or the coreml-onnx-model. In that case you'd have to be patient, until our beloved developer is back on his feet. Do you have the same issue with the other enhancers? I'm getting good results with GPEN, but performance is roughly half of GFPGAN.

Regarding performance: Which graphics card are you using? With roop-unleashed, espescially the amount of VRAM is important. Most performance issues that have been posted here, are related to exhausting all of the available VRAM as a result of running too many threads. When that happens, performance will drop drastically. I can't see exactly, how many threads you are running in your screenshot (lower right corner), but it seems to be double digits (10+). With my RTX3060 (12GB VRAM) and GFPGAN, 3 threads was the maximum if I remember correctly. With GPEN as enhancer, 2 threads is the maximum for the RTX3060, resulting in 9-10GB total VRAM usage. With 3 threads, it would use more than the available 12GB, resulting in poor performance. So try a lower thread count and watch your VRAM usage in the task manager.

**Comment by bennyc42069** [comment]
how do i get roop unleashed on my mac? Having major issues with my m1 getting it on there. Novice here. Can anyone link to a video or the mac code copy and paste?

**Comment by Oil3** [comment]
> how do i get roop unleashed on my mac? Having major issues with my m1 getting it on there. Novice here. Can anyone link to a video or the mac code copy and paste?

how far did u go?
it(s not complicated  but its complex. 
Progratmers worldwide seem to enjoy buildin the most complex stuff for the most simple things. 
Think about curl and how ridiculous it is to just download a ile.
"vi" the minimalist text editor. I mean. THey don't tell you how to close the thing and it uses obscure shortcuts.
Mhmm ever tried to make X11 work?
Oh yes, and thet automatically assume everybody already know what to do and when you ask how to install they go just "build and make like usual". Or they invent yet another package manager. Conda? Comon why? I think they procrastinate and develop stuf tht alreadty exist in lieu of working on their real project. I personally do a lot o things like that. 
Lol. But without them/us, we'd still be on pen and paper.

ANswering your question felt wiki-worhty
https://github.com/C0untFloyd/roop-unleashed/wiki/Installation-:-MacOS
let me know what else, it will help to finish the draft,
basically do the steps and tell me the error message if/when it bugs

**Comment by bennyc42069** [comment]
This is amazing. I want to thank you for all your hard work and your
transparency in helping me better understand everything. As I move forward
in this email I will try my best to explain everything fully. I will also
be attaching images (screen shots) labeled FIGX so you can fully see
everything.

Ok I have followed these steps you have provided. It seems that I had
already done everything correct before you reached out to me, EXCEPT STEP
3. That was my previous wall and I have moved forward thanks to you.

That being said, I have since completed step 3 and received "Fig1".

Beyond that, I said fuck it lets keep trucking I hope that doesnt stop me,
on to step 4. When trying to complete step 4, I received  "Fig2".

I have 2 macs. My mac mini, and my macbook pro. It seems that for some
reason, It keeps trying to access a different mac? My mac mini is the one
that I'm operating on and hope to run roop unleashed on. It's also where
I'm doing the terminal commands. My mac mini is called "jonathan's macbook
pro" and my other computer is called "Jonathans Macbook pro 2" "Fig3". For
some reason in the terminal it says "Fig4". Does this mean its trying to
access a different home folder? My roop is in the correct home folder as
stated in STEP 1 of your directions "Fig5". But it seems the terminal path
is trying to grab it from a diff directory. I have also asked gpt4 whats up
and it told me some things that im nervous on doing, because i dont want my
whole computer to get confused as fuck and not know where anything is, if i
change the name of it. I just need to make terminal go the correct
direction. "Fig 6&7".

Thanks for all your help.

**Comment by bennyc42069** [comment]
wait, i believe i found the problem. what next? the original issue was in
step 4. "python3 ~/roop-unleashed/run.py" should really says "python3
~/roop-unleashed-main/run.py"

but now this torch thing

**Comment by bennyc42069** [comment]
Interesting! I'm always up for a challenge myself.

Here are all the screen shots again.

**Comment by Oil3** [comment]
@bennyc42069 can you plz try putting the screenshot via the website and not email?

**Comment by bennyc42069** [comment]
Hey, I actually figured it out! Basically all the figures I gave you are now irrelevant! Once I copy and pasted the requirments, one by one, it worked. This tech is awesome! Trying to get the hang of face sets etc though. Are face sets basically, multiple faces on one individual give the code more ability to render correctly?

**Comment by Oil3** [comment]
Much better you fixed it yourself because, lol, there will be more!

from code, it takes the mean average of all faces. 
For my experience, more photos better in the set, better.
Maybe in the future we'll have angles taken in account a la 3D and not just the mean
